FTBFS due to mismatch during inlining

**Describe the bug**
FTBFS due to inlining.

**To Reproduce**
Steps to reproduce the behavior:

1. Build package from [AUR](https://aur.archlinux.org/packages/dragonflydb). EDIT: With the commit changed to point at `0.14.0`.
2. See below CMake output:

```sh
[297/398] Building CXX object src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o
FAILED: src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o
/usr/bin/g++ -DBOOST_ASIO_SEPARATE_COMPILATION -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_BEAST_SEPARATE_COMPILATION -DBOOST_CONTEXT_DYN_LINK -DBOOST_CONTEXT_NO_LIB -DBOOST_FIBER_DYN_LINK -DBOOST_FIBER_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DGLOG_CUSTOM_PREFIX_SUPPORT -DUSE_ZMALLOC_MI -I/build/dragonflydb/src/dragonflydb -I/build/dragonflydb/src/dragonflydb/genfiles -I/build/dragonflydb/src/dragonflydb/src -I/build/dragonflydb/src/dragonflydb/helio -I/build/dragonflydb/src/dragonflydb/build-opt/_deps/glog-build -I/build/dragonflydb/src/dragonflydb/build-opt/_deps/glog-src/src -I/build/dragonflydb/src/dragonflydb/build-opt/_deps/abseil_cpp-src -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/xxhash/include -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/mimalloc/include -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/lua/include -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/jsoncons/include -Wall -Wextra -g -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -fno-omit-frame-pointer -Wno-unused-parameter -march=sandybridge -mtune=skylake -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -flto=auto -std=c++20 -DHAS_RAWMEMCHR -fdiagnostics-color=always  -O3 -DNDEBUG  -std=gnu++17 -MD -MT src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o -MF src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o.d -o src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o -c /build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc
In file included from /build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:13:
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h: In function ‘dfly::detail::simd_variant1_pack(char const*, char const*, unsigned char*)’:
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8(long long __vector(2), long long __vector(2))’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
/build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:65:27: note: called from here
   65 |     val = _mm_shuffle_epi8(val, control);
      |           ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8(long long __vector(2), long long __vector(2))’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
/build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:65:27: note: called from here
   65 |     val = _mm_shuffle_epi8(val, control);
      |           ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8(long long __vector(2), long long __vector(2))’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
/build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:65:27: note: called from here
   65 |     val = _mm_shuffle_epi8(val, control);
      |           ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
[298/398] Building CXX object src/core/CMakeFiles/dfly_core.dir/tx_queue.cc.o
```

**Environment (please complete the following information):**
 - OS: Arch Linux
 - Kernel: 6.1.8
 - Containerized?: Bare metal building system package in a systemd-nspawn container (has worked for the past few releases, and the error removes this as a cause).
 - Dragonfly Version: 0.14.0

It looks like it still fails when that line is commented out:

```sh
[301/398] Building CXX object src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o
FAILED: src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o
/usr/bin/g++ -DBOOST_ASIO_SEPARATE_COMPILATION -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_BEAST_SEPARATE_COMPILATION -DBOOST_CONTEXT_DYN_LINK -DBOOST_CONTEXT_NO_LIB -DBOOST_FIBER_DYN_LINK -DBOOST_FIBER_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DGLOG_CUSTOM_PREFIX_SUPPORT -DUSE_ZMALLOC_MI -I/build/dragonflydb/src/dragonflydb -I/build/dragonflydb/src/dragonflydb/genfiles -I/build/dragonflydb/src/dragonflydb/src -I/build/dragonflydb/src/dragonflydb/helio -I/build/dragonflydb/src/dragonflydb/build-opt/_deps/glog-build -I/build/dragonflydb/src/dragonflydb/build-opt/_deps/glog-src/src -I/build/dragonflydb/src/dragonflydb/build-opt/_deps/abseil_cpp-src -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/xxhash/include -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/mimalloc/include -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/lua/include -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/jsoncons/include -Wall -Wextra -g -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -fno-omit-frame-pointer -Wno-unused-parameter -march=sandybridge -mtune=skylake -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -g -ffile-prefix-map=/build/dragonflydb/src=/usr/src/debug/dragonflydb -flto=auto -std=c++20 -DHAS_RAWMEMCHR -fdiagnostics-color=always  -O3 -DNDEBUG  -std=gnu++17 -MD -MT src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o -MF src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o.d -o src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o -c /build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc
In file included from /build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:13:
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h: In function ‘dfly::detail::simd_variant1_pack(char const*, char const*, unsigned char*)’:
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8(long long __vector(2), long long __vector(2))’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
/build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:65:27: note: called from here
   65 |     val = _mm_shuffle_epi8(val, control);
      |           ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8(long long __vector(2), long long __vector(2))’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
/build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:65:27: note: called from here
   65 |     val = _mm_shuffle_epi8(val, control);
      |           ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8(long long __vector(2), long long __vector(2))’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
/build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:65:27: note: called from here
   65 |     val = _mm_shuffle_epi8(val, control);
      |           ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
[302/398] Building CXX object src/core/CMakeFiles/dfly_core.dir/mi_memory_resource.cc.o
```

I see that in the compile command you used these overrides "
-march=sandybridge -mtune=skylake *-march=x86-64 -mtune=generic*" - you
override the arch with x86-64 and it probably does not have the required
instruction set.
Try adding "-mavx" or restoring the original sandy bridge architecture.

On Mon, Jan 30, 2023 at 9:31 AM George Rawlinson ***@***.***>
wrote:

> It looks like it still fails when that line is commented out:
>
> [301/398] Building CXX object src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o
>
> FAILED: src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o
>
> /usr/bin/g++ -DBOOST_ASIO_SEPARATE_COMPILATION -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_BEAST_SEPARATE_COMPILATION -DBOOST_CONTEXT_DYN_LINK -DBOOST_CONTEXT_NO_LIB -DBOOST_FIBER_DYN_LINK -DBOOST_FIBER_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DGLOG_CUSTOM_PREFIX_SUPPORT -DUSE_ZMALLOC_MI -I/build/dragonflydb/src/dragonflydb -I/build/dragonflydb/src/dragonflydb/genfiles -I/build/dragonflydb/src/dragonflydb/src -I/build/dragonflydb/src/dragonflydb/helio -I/build/dragonflydb/src/dragonflydb/build-opt/_deps/glog-build -I/build/dragonflydb/src/dragonflydb/build-opt/_deps/glog-src/src -I/build/dragonflydb/src/dragonflydb/build-opt/_deps/abseil_cpp-src -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/xxhash/include -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/mimalloc/include -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/lua/include -isystem /build/dragonflydb/src/dragonflydb/build-opt/third_party/libs/jsoncons/include -Wall -Wextra -g -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -fno-omit-frame-pointer -Wno-unused-parameter -march=sandybridge -mtune=skylake -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -g -ffile-prefix-map=/build/dragonflydb/src=/usr/src/debug/dragonflydb -flto=auto -std=c++20 -DHAS_RAWMEMCHR -fdiagnostics-color=always  -O3 -DNDEBUG  -std=gnu++17 -MD -MT src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o -MF src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o.d -o src/core/CMakeFiles/dfly_core.dir/detail/bitpacking.cc.o -c /build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc
>
> In file included from /build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:13:
>
> /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h: In function ‘dfly::detail::simd_variant1_pack(char const*, char const*, unsigned char*)’:
>
> /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8(long long __vector(2), long long __vector(2))’: target specific option mismatch
>
>   136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
>
>       | ^~~~~~~~~~~~~~~~
>
> /build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:65:27: note: called from here
>
>    65 |     val = _mm_shuffle_epi8(val, control);
>
>       |           ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
>
> /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8(long long __vector(2), long long __vector(2))’: target specific option mismatch
>
>   136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
>
>       | ^~~~~~~~~~~~~~~~
>
> /build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:65:27: note: called from here
>
>    65 |     val = _mm_shuffle_epi8(val, control);
>
>       |           ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
>
> /usr/lib/gcc/x86_64-pc-linux-gnu/12.2.1/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8(long long __vector(2), long long __vector(2))’: target specific option mismatch
>
>   136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
>
>       | ^~~~~~~~~~~~~~~~
>
> /build/dragonflydb/src/dragonflydb/src/core/detail/bitpacking.cc:65:27: note: called from here
>
>    65 |     val = _mm_shuffle_epi8(val, control);
>
>       |           ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
>
> [302/398] Building CXX object src/core/CMakeFiles/dfly_core.dir/mi_memory_resource.cc.o
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dragonflydb/dragonfly/issues/732#issuecomment-1408118933>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4BFCE7RP4RR7HGQFK5XNLWU5U37ANCNFSM6AAAAAAUKQOK5U>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

--- a/.github/workflows/daily-builds.yml
+++ b/.github/workflows/daily-builds.yml
@@ -1,0 +1,60 @@
+name: daily-builds
+
+on:
+  # TODO: Remove this. This is for testing
+  schedule:
+    - cron: '0 6 * * *' # run at 6 AM UTC
+  workflow_dispatch:
+
+jobs:
+  build:
+    # The CMake configure and build commands are platform agnostic and should work equally
+    # well on Windows or Mac.  You can convert this to a matrix build if you need
+    # cross-platform coverage.
+    # See: https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/managing-complex-workflows#using-a-build-matrix
+    runs-on: ubuntu-latest
+    name: Build with ${{ matrix.flags }} flags
+    strategy:
+      matrix:
+        # Build with these flags
+        flags: ["-DMARCH_OPT=-march=x86_64"]
+    timeout-minutes: 30
+
+    container:
+      image: ghcr.io/romange/alpine-dev:latest
+      credentials:
+        username: ${{ github.repository_owner }}
+        password: ${{ secrets.GITHUB_TOKEN }}
+
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: true
+      - name: Install dependencies
+        run: |
+          uname -a
+          cmake --version
+          mkdir -p ${{github.workspace}}/build
+      - name: Cache build deps
+        id: cache-deps
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.ccache
+            ${{github.workspace}}/build/_deps
+          key: ${{ runner.os }}-deps-${{ github.base_ref }}-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-deps-${{ github.base_ref }}-
+
+      - name: Configure CMake
+        # Configure CMake in a 'build' subdirectory. `CMAKE_BUILD_TYPE` is only required if you are using a single-configuration generator such as make.
+        # See https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html?highlight=cmake_build_type
+        run: |
+          cmake -B ${{github.workspace}}/build -DCMAKE_BUILD_TYPE=Debug -GNinja -DCMAKE_CXX_COMPILER_LAUNCHER=ccache ${{ matrix.flags }}
+          cd ${{github.workspace}}/build && pwd
+          du -hcs _deps/
+      - name: Build
+        run: |
+          cd ${{github.workspace}}/build
+          ninja src/all
+          ccache --show-stats

